rfcs/0078-configuration-file-generators.md

```diff
@@ -0,0 +1,146 @@
+following properties:
+* unlike, say, upstream `systemd` unit files we typically do not
+  reuse-and-override upstream configs, but generate our own (both in NixOS and
+  other service collections)
+* the configs for a piece of software often have few dependencies on the
+  general environment
+* it can be defined as a purely functional library, simplifying integration
+
+We aim to reduce duplication of effort and code while also having a high level
+of inspection and tweaking possibilities without needing to patch module
+source.
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC proposal builds upon RFCs#42 defining a configuration file abstraction
+approach.
+
+A subtree is established in the Nixpkgs repository serving as a library of
+functions implementing program configuration file generation. .As the design
+uses a module system, putting configuration generators next to packages is also
+not a perfect solution; also a configuration generator might end up generating
+a set of config for a combination of packages.
+
+Each such generator takes as an input a NixOS module system based «subtree»,
+e.g. the values typically bound to `cfg` variable in typical modules.
+
+The output of each such function is an attribute set. The keys are relative
+file names of configuration files, and the values are attribute sets of RFCs#42
+settings abstraction and serialiser.
+
+Along with the function the corresponding type specifications for input and
+output modules are provided.
+
+In some cases we only provide low-level overridable default configuration. In
+this case the input may have the type that is always an empty attribute set,
+`{}`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+A minimalistic silly example is:
+
+Input:
+```
+  {
+     port = 1234;
+     nodeName = "localhost";
+     content = "Hello";
+  }
+```
+
+Output:
+```
+  {
+    "subdir/listen.conf" = {
+      serialiser = toJSON;
+      settings = {
+        listenPort = "1234";
+        serverGreet = "I am localhost";
+      };
+      type = …;
+    };
+    "subdir/content.conf" = {
+      serialiser = toINI;
+      settings = {
+        greeting = {
+          greeter = "localhost";
+        };
+        payload = {
+          "I say to you: ‘Hello’";
+        };
+      };
+      type = …;
+    };
+  }
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The proposed approach increases the spreading of the related code, from package
+and service module to package and configuration generator and module.
+
+The proposed refactoring incurs a cost in terms of implementation effort across
+NixOS, and likely to create a medium-term situation of partial migration.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Do nothing, continuing with the current code duplication.
+
+Same, but put configuration generators closer to packages.
+
+Many proposed solutions based on a significant rework of the module system.
```

Comment:
@blaggacao, I'm pretty sure it means...
```suggestion
There also have been many solutions proposed based on a significant rework of the module system.
```

Reply:
So it would be probably something around:

> There also have been many proposals that factor out configuration generation based on a significant rework of the module system, such as ...

Reply:
It does. It is a non-normative section, I confirm that I have not ignored this direction, and describe it in a way that emphasises that a lot of work would be needed, taking a lot of time. Describing any of the options well would probably require a full slide deck. Like that talk about modules at NixCon.